### PR TITLE
Support  the new Microsoft Quick input method for Chinese Traditional (Hong Kong)

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
Windows 10 introduces a newer version of the Microsoft Quick input method for Chinese Traditional (Hong Kong). these new version is all UI automation, integrated into the  windowsinternal composableshell_experiences textinput_inputapp.
Users expect Automatically reading all candidates in a list, including their number, and optional short character descriptions.

 ### Description of how this pull request fixes the issue:
Ensure that these UI automation candiate items and lists get the correct behaviour in NVDA by making use of our existing candidate behaviour NVDAObjects.

### Testing performed:
* In Windows 10 settings:
 * Add the chinese Traditional (Hong Kong) language
 * Add the Microsoft Quick input method as a keyboard to this language if not there already.
* In Notepad or another text editor:
 * Switch to chinese Traditional (Hong Kong) language with alt+shift
 * Ensure that you are in native input mode by tapping leftShift+rightShift until you are.
 * Type h then i. A candidate list should appear and NVDA should read it all. Moving the up and down arrows should speak the currently selected item.
 * Press 5 to choose the 5th candidate. A new Associated phrases list should appear and NVDA should automatically read all candidates.
 * Press shift+1 to choose the first assocated phrase and NVDA should announce that these symbols are typed into the document.
 
### Known Issues
None

### change log
New features:
- Added support for the new chinese Traditional Quick Input method in Windows 10.

### CODE REVIEW

**Comment by michaelDCurran** [comment]
@josephsl  I have asked for your review on this pr as it directly touches a lot of your work in windowsinternal_composableshell_experiences_textinput_inputapp.py.
I hope this does not majorly conflict with any planned work you have -- but the Hong Kong community is quite interested in seeing this support very soon, in order that they can again enter characters with NvDA on Windows 10.

**Comment by lijianwei2019** [comment]
@larry801

**Comment by michaelDCurran** [comment]
try build:
https://ci.appveyor.com/api/buildjobs/gaoaplkrqdd5mkte/artifacts/output%2Fnvda_snapshot_try-microsoftQuickInput-20835%2C1388e4d4.exe

**Comment by larry801** [comment]
Tested the snapshot on Windows 10  19042 20207 
Candidate and description can be announced.
When candidate has only one character,  there is a long form description follows short form description, which is quite annoying.

Speaking [LangChangeCommand ('zh_CN'), '(你们的你)、1. 称对方，多称指一个人，有时也指称若干人：你厂。你方。 2. 泛指任何人：你死我活。', EndUtteranceCommand()]

**Comment by michaelDCurran** [comment]
Did this behavior happen with the old Microsoft Quick input method that 
NVDA supported?

This pr is only to ensure that the New Quick Input Method on Windows 10 
acts the same as the old one.

If this issue is separate to this work, please file a separate issue.


If this issue is new to this pr, please provide exact steps to 
reproduce, including what characters to type, and what you expect.

**Comment by larry801** [comment]
Yes it happens on old Microsoft Quick input method. I will file another issue instead

**Comment by larry801** [comment]
Found another issue about New Microsoft Quick

press h and y then use down and right arrow keys to navigate around candidates 
nvda keeps reading blank

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 1]
```diff
@@ -11,19 +11,109 @@
...
+	The UIAutomation-based IME candidate UI (such as for  the modern Chinese Microsoft Quick input).
+	This class ensures NVDA is notified of the first selected item when the UI is shown.
+	"""
+
+	def event_show(self):
+		# The IME candidate UI is shown.
+		# Report the current candidates page and the currently selected item.
+		# Sometimes UIA does not fire an elementSelected event when it is first opened,
+		# Therefore we must fake it here.
+		if (
```
Comment (larry801) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:36]:
```suggestion
		if (self.UIAAutomationId == "IME_Prediction_Window"):
			candidateItem = self.firstChild
			eventHandler.queueEvent("UIA_elementSelected", candidateItem)
		elif (

```
New Microsoft Pinyin IME in Windows 10 20H1 emits similar events with New Microsoft Quick. New Microsoft Pinyin IME input method can also benefit from this pr with some modifications.

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 2]
```diff
@@ -130,3 +233,19 @@ def event_nameChange(self, obj, nextHandler):
...
+		if isinstance(obj, UIA):
+			if obj.role == controlTypes.ROLE_LISTITEM and (
+				(
+					obj.parent.UIAAutomationId in (
+						"ExpandedCandidateList",
+						"TEMPLATE_PART_AdaptiveSuggestionList",
+					)
+					and obj.parent.parent.UIAAutomationId == "IME_Candidate_Window"
+				)
+				or obj.parent.UIAAutomationId == "IME_Candidate_Window"
```
Comment (larry801) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:247]:
```suggestion
				or obj.parent.UIAAutomationId in ("IME_Candidate_Window", "IME_Prediction_Window")

```
In Microsoft pinyin automation id is "IME_Prediction_Window"

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 3]
```diff
@@ -130,3 +233,19 @@ def event_nameChange(self, obj, nextHandler):
...
+			):
+				clsList.insert(0, ImeCandidateItem)
+			elif obj.role == controlTypes.ROLE_PANE and obj.UIAAutomationId == "IME_Candidate_Window":
```
Comment (larry801) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:250]:
```suggestion
			elif obj.role == controlTypes.ROLE_PANE and obj.UIAAutomationId in ("IME_Candidate_Window", "IME_Prediction_Window"):

```

**Comment by michaelDCurran** [comment]
@larry801  are you saying that if I commit your suggestions here the new Pinyin input method will also work? or are there other changes needed as well? If so then perhaps we should leave that for another pr. But if your suggested changes are all that is needed at a minimum then I'll be happy to take them here.

**Comment by larry801** [comment]
@michaelDCurran  Yes the three changes are all that is needed at a minimum.

**Comment by josephsl** [comment]
Hi, note that this change may affect Japanese and Korean modern IME. CC @Nishomotz for confirmation. Thanks.

File: source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py [hunk 4]
```diff
@@ -11,19 +11,112 @@
...
+			return
+		api.setNavigatorObject(self)
+		speech.cancelSpeech()
+		# Report the entire current page of candidate items if it is newly shown  or it has changed.
+		if config.conf["inputComposition"]["autoReportAllCandidates"]:
+			oldText = getattr(self.appModule, '_lastImeCandidateVisibleText', '')
+			newText = self.visibleCandidateItemsText
+			if not isinstance(oldNav, ImeCandidateItem) or newText != oldText:
+				print(f"oldText: {oldText}")
+				print(f"newText: {newText}")
```
Comment (feerrenrut) [inline source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:102]: I think this might need to be removed.

**Review by feerrenrut — DISMISSED** [review]
Generally this looks fine to me, and I think there is plenty of time for this to get testing.

**Comment by josephsl** [comment]
Hi mick and everyone,

I must say this PR is much better than the version I have implemented in Windows 10 App Essentials, as it also supports Hanja entry in 21H1 (Korean). Once merged into master, I'll "backport" this to NVDA 2020.1 environment via WinTenApps so folks can preview this change.

Thanks.

**Comment by josephsl** [comment]
Follow-up: this PR may also fix #10093 provided that @Nishimotz can confirm it. Thanks.

**Comment by josephsl** [comment]
Hi,

Major regression found: this PR breaks hardware keyboard input suggestions feature i.e. whenever suggestions change, NVDA will not announce them. Part of this has to do with no handling for name change event in IME candidate item class or not comparing hardware input suggestion items.

Thanks.

**Comment by josephsl** [comment]
Hi,

Never mind - hardware suggestion announcement is working - I have enabled selective UIA event registration, so it's my settings.

Thanks.

**Comment by feerrenrut** [comment]
If I'm not wrong, no issues have been found with this. If the debug logging is removed this should be ready to merge.

**Pull request merged**